Call `didSendRequestPart` at the right time

FranzBusch opened the pull request:
# Motivation
Right now, we call `didSendRequestPart` after passing the write to the executor. However, this does not mean that the write hit the socket. To implement proper backpressure using the delegate, we should only call this method once the write was successful.

# Modification
Pass a promise to the actual channel write and only call the delegate once that promise succeeds.

# Result
The delegate method `didSendRequestPart` is only called after the write was successful.

dnadoba commented:
@swift-nio-bot add to allowlist

Lukasa reviewed (changes requested):
There needs to be much clearer auditing about who is responsible for succeeding this promise. A bunch of these calls return without either storing the promise or passing it on, and in those cases we risk leaking the promise without succeeding it. That all needs to be updated.

FranzBusch commented:
@Lukasa Thanks for your review. You absolutely write I just dropped the promise in a couple of places. I audited the code paths again and made sure to fail the promise in the other branches and added some tests.

The only thing that I wasn't really sure about is what errors to return in some places. Looking forward to suggestions!

Lukasa commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift:
```diff
@@ -281,17 +281,18 @@ final class HTTP2ClientRequestHandler: ChannelDuplexHandler {
             // the request has been popped by the state machine or the ChannelHandler has been
             // removed from the Channel pipeline. This is a normal threading issue, noone has
             // screwed up.
+            promise?.fail(ChannelError.eof)
```
Where you've used `ChannelError.eof`, I don't think that's right: I think we want `ioOnClosedChannel`.

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -965,6 +966,8 @@ public struct HTTPClientError: Error, Equatable, CustomStringConvertible {
     /// Remote connection was closed unexpectedly.
     public static let remoteConnectionClosed = HTTPClientError(code: .remoteConnectionClosed)
+    /// Request failed.
+    public static let failed = HTTPClientError(code: .failed)
```
Can we call this `requestFailed` instead of just `failed`?

FranzBusch replied:
I was aligning it with the naming of `cancelled` which is documented as a cancelled request. What do you prefer?

Lukasa commented on Sources/AsyncHTTPClient/RequestBag.swift:
```diff
@@ -154,8 +154,11 @@ final class RequestBag<Delegate: HTTPClientResponseDelegate> {
-            self.delegate.didSendRequestPart(task: self.task, part)
+            let promise = self.task.eventLoop.makePromise(of: Void.self)
+            promise.futureResult.whenSuccess {
+                self.delegate.didSendRequestPart(task: self.task, part)
```
Should we report an error to the delegate in the failure case?

FranzBusch replied:
Was thinking about that as well. Open to input here. @weissi For your backpressure implementation would this be valuable?

fabianfett replied:
We should make sure the error delegate is only called once.

Lukasa reviewed (approved):
LGTM but I'd like to see @fabianfett or @dnadoba review.

FranzBusch commented:
The API breakage is failing because of the two new protocol requirements. However, I added default implementations in an extension so this is not a breaking change.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -290,6 +291,7 @@ struct HTTPRequestStateMachine {
             if let expected = expectedBodyLength, sentBodyBytes + part.readableBytes > expected {
                 let error = HTTPClientError.bodyLengthMismatch
                 self.state = .failed(error)
+                promise?.fail(error)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -303,9 +305,10 @@ struct HTTPRequestStateMachine {
+            return .sendBodyPart(part, promise)
 
         case .failed:
+            promise?.fail(HTTPClientError.failed)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -336,35 +340,40 @@ struct HTTPRequestStateMachine {
             if let expected = expectedBodyLength, expected != sentBodyBytes {
                 let error = HTTPClientError.bodyLengthMismatch
                 self.state = .failed(error)
+                promise?.fail(error)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -336,35 +340,40 @@ struct HTTPRequestStateMachine {
             if let expected = expectedBodyLength, expected != sentBodyBytes {
                 let error = HTTPClientError.bodyLengthMismatch
                 self.state = .failed(error)
+                promise?.fail(error)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -336,35 +340,40 @@ struct HTTPRequestStateMachine {
             if let expected = expectedBodyLength, expected != sentBodyBytes {
                 let error = HTTPClientError.bodyLengthMismatch
                 self.state = .failed(error)
+                promise?.fail(error)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -336,35 +340,40 @@ struct HTTPRequestStateMachine {
             }
 
             self.state = .finished
+            promise?.succeed(())
```
See above. Succeed promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -336,35 +340,40 @@ struct HTTPRequestStateMachine {
             return .succeedRequest(.sendRequestEnd, .init())
 
         case .failed:
+            promise?.fail(HTTPClientError.failed)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -377,6 +386,7 @@ struct HTTPRequestStateMachine {
 
             // We may still receive something, here because of potential race conditions with the
             // producing thread.
+            promise?.fail(ChannelError.eof)
```
See above. Fail promise outside of state machine lock.

fabianfett commented on Sources/AsyncHTTPClient/HTTPHandler.swift:
```diff
@@ -431,12 +431,27 @@ public protocol HTTPClientResponseDelegate: AnyObject {
     ///     - part: Request body `Part`.
     func didSendRequestPart(task: HTTPClient.Task<Response>, _ part: IOData)
 
+    /// Called when a part of the request body failed to sent. Could be called zero or more times.
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - part: Request body `Part`.
+    ///     - error: Error that occured during sending the request part.
+    func didFailSendingRequestPart(task: HTTPClient.Task<Response>, _ part: IOData, error: Error)
+
```
Why do we need to extend the public API here?

fabianfett commented on Sources/AsyncHTTPClient/HTTPHandler.swift:
```diff
@@ -431,12 +431,27 @@ public protocol HTTPClientResponseDelegate: AnyObject {
     ///     - task: Current request context.
     func didSendRequest(task: HTTPClient.Task<Response>)
 
+    /// Called when the request failed to sent. Will be called once.
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - error: Error that occured during sending the request.
+    func didFailSendingRequest(task: HTTPClient.Task<Response>, error: Error)
+
```
Why do we need to extend the public API here?

fabianfett commented on Sources/AsyncHTTPClient/RequestBag.swift:
```diff
@@ -154,8 +154,16 @@ final class RequestBag<Delegate: HTTPClientResponseDelegate> {
+                case .success:
+                    self.delegate.didSendRequestPart(task: self.task, part)
+                case .failure(let error):
+                    self.delegate.didFailSendingRequestPart(task: self.task, part, error: error)
```
I think, we should not change the old API at the moment and should just ignore this case. The error will be reported to the user anyway.

fabianfett commented on Sources/AsyncHTTPClient/RequestBag.swift:
```diff
@@ -168,11 +176,18 @@ final class RequestBag<Delegate: HTTPClientResponseDelegate> {
+                case .success:
+                    self.delegate.didSendRequest(task: self.task)
+                case .failure(let error):
+                    self.delegate.didFailSendingRequest(task: self.task, error: error)
```
Same as above. Don't extend public API here.

fabianfett reviewed (commented):
Thanks so much for driving this. Left a bunch of comments.

weissi commented on Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift:
```diff
@@ -274,6 +274,7 @@ struct HTTPRequestStateMachine {
             // won't be interested. We expect that the producer has been informed to pause
             // producing.
             assert(producerState == .paused)
+            promise?.fail(HTTPClientError.requestStreamCancelled)
```
+1, yeah we can't do that. That'll "invite" arbitrary code right into the state machine's lock and makes the state machine have implicit side effects.

fabianfett reviewed (commented):
Looks much better now. However we should change one more critical piece.

fabianfett commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -937,6 +937,8 @@ public struct HTTPClientError: Error, Equatable, CustomStringConvertible {
         case getConnectionFromPoolTimeout
         case deadlineExceeded
         case httpEndReceivedAfterHeadWith1xx
+        case requestBodyStreamCancelled
```
There already is `requestStreamCancelled`, maybe we can reuse that?

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -269,8 +269,9 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
             }
 
             oldRequest.succeedRequest(buffer)
+            promise?.succeed(())
```
this promise should be passed to:

```swift
            case .sendRequestEnd:
                context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
```

this will be more obvious once we attach the promise to `sendRequestEnd` and not to `succeedRequest`. In this case we should also defer the sending of `oldRequest.succeedRequest(buffer)` until the promise has fired, since we may run into ordering issues otherwise.

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -269,8 +269,9 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
+            promise?.succeed(())
 
-        case .failRequest(let error, let finalAction):
+        case .failRequest(let error, let finalAction, let promise):
```
```suggestion
        case .failRequest(let error, let finalAction, let writePromise):
```

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -288,6 +289,10 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
             }
 
             oldRequest.fail(error)
+            promise?.fail(error)
```
```suggestion
            writePromise?.fail(error)
```

fabianfett commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -355,27 +360,28 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
             // the request has been popped by the state machine or the ChannelHandler has been
             // removed from the Channel pipeline. This is a normal threading issue, noone has
             // screwed up.
+            promise?.fail(ChannelError.eof)
```
```suggestion
            promise?.fail(HTTPClientError.requestStreamCancelled)
```

fabianfett commented on Sources/AsyncHTTPClient/RequestBag.swift:
```diff
@@ -168,11 +171,13 @@ final class RequestBag<Delegate: HTTPClientResponseDelegate> {
-            promise?.succeed(())
-
-            self.delegate.didSendRequest(task: self.task)
+        case .forwardStreamFinished(let writer, let writerPromise):
+            let promise = self.task.eventLoop.makePromise(of: Void.self)
+            promise.futureResult.whenSuccess {
+                self.delegate.didSendRequest(task: self.task)
+            }
+            writer.finishRequestBodyStream(self, promise: promise)
+            writerPromise?.succeed(())
```
Why do we create a new promise? Why is the writePromise just succeeded? This looks odd. This needs at least a comment.

FranzBusch replied:
I thought that we need a new promise here but I just tested it again and with using the same promise and it worked.

FranzBusch commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -258,36 +258,52 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
+                    oldRequest.succeedRequest(buffer)
+                }
+
+                context.close(promise: writePromise)
```
@fabianfett Is it okay to pass the `writePromise` to `close` here?

FranzBusch commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:
```diff
@@ -258,36 +258,52 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
-                context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
-            case .informConnectionIsIdle:
+            case .close(let writePromise):
+                context.close(promise: writePromise)
```
Same question here @fabianfett ?

fabianfett reviewed (commented):
And another round!

fabianfett reviewed (commented):
Nearly done!

FranzBusch commented on Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ConnectionStateMachine.swift:
```diff
@@ -390,41 +408,57 @@ extension HTTP1ConnectionStateMachine.State {
+            // otherwise we fail with .informConnectionIsIdle
+            case (.inRequest(_, _), .none):
+                self = .idle
+                return .failRequest(error, .informConnectionIsIdle(nil))
```
Changed it back

fabianfett reviewed (approved):
Thank you! This has been a ton of work!
